src/simpleppl.jl

```diff
@@ -0,0 +1,749 @@
+const VectorVariate = ArrayVariate{1}
+const MatrixVariate = ArrayVariate{2}
+
+
+#################### Distribution Types ####################
+
+const DistributionStruct =
+    Union{Distribution,Array{UnivariateDistribution},Array{MultivariateDistribution}}
+
+#################### Null Distribution ####################
+
+struct NullUnivariateDistribution <: UnivariateDistribution{ValueSupport} end
+
+
+#################### Dependent Types ####################
+
+mutable struct ScalarLogical <: ScalarVariate
+    value::Float64
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+end
+
+mutable struct ArrayLogical{N} <: ArrayVariate{N}
+    value::Array{Float64,N}
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+end
+
+mutable struct ScalarStochastic <: ScalarVariate
+    value::Float64
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+    distr::UnivariateDistribution
+end
+
+mutable struct ArrayStochastic{N} <: ArrayVariate{N}
+    value::Array{Float64,N}
+    symbol::Symbol
+    monitor::Vector{Int}
+    eval::Function
+    sources::Vector{Symbol}
+    targets::Vector{Symbol}
+    distr::DistributionStruct
+end
+
+const AbstractLogical = Union{ScalarLogical,ArrayLogical}
+const AbstractStochastic = Union{ScalarStochastic,ArrayStochastic}
+const AbstractDependent = Union{AbstractLogical,AbstractStochastic}
+
+#################### Variate ####################
+
+#################### Conversions ####################
+
+Base.convert(::Type{Bool}, v::ScalarVariate) = convert(Bool, v.value)
+Base.convert(::Type{T}, v::ScalarVariate) where {T<:Integer} = convert(T, v.value)
+Base.convert(::Type{T}, v::ScalarVariate) where {T<:AbstractFloat} = convert(T, v.value)
+Base.AbstractFloat(v::ScalarVariate) = convert(Float64, v)
+Base.Float64(v::ScalarVariate) = convert(Float64, v)
+
+Base.convert(::Type{Matrix}, v::MatrixVariate) = v.value
+Base.convert(::Type{Vector}, v::VectorVariate) = v.value
+Base.convert(
+    ::Union{Type{Array{T}},Type{Array{T,N}}},
+    v::ArrayVariate{N},
+) where {T<:Real,N} = convert(Array{T,N}, v.value)
+
+Base.unsafe_convert(::Type{Ptr{Float64}}, v::ArrayVariate) = pointer(v.value)
+
+
+macro promote_scalarvariate(V)
+    quote
+        Base.promote_rule(::Type{$(esc(V))}, ::Type{T}) where {T<:Real} = Float64
+    end
+end
+
+
+#################### Base Functions ####################
+
+Base.size(v::AbstractVariate) = size(v.value)
+
+Base.stride(v::ArrayVariate, k::Int) = stride(v.value, k)
+
+
+#################### Indexing ####################
+
+Base.getindex(v::ScalarVariate, ind::Int) = v.value[ind]
+
+Base.getindex(v::ScalarVariate, inds::Union{StepRange{Int,Int},Vector{Int}}) =
+    Float64[v[i] for i in inds]
+
+Base.getindex(v::ArrayVariate, inds::Int...) = getindex(v.value, inds...)
```

Comment:
I guess if you define the `IndexStyle` as `IndexLinear` (might not be possible) then this should not be needed either.